vlasovsolver/cpu_trans_pencils.hpp

```diff
@@ -97,6 +106,69 @@ struct setOfPencils {
       path.push_back(pathIn);
    }
 
+   void binPencils(const std::unordered_set<CellID>& targetCells) {
+      bins.resize(N);
+
+      // Loop over pencils
+      for (uint i = 0; i < N; ++i) {
+         bins[i] = i;
+         binsCells[i] = {};
+
+         for (auto id = ids.begin() + idsStart[i]; id < ids.begin() + idsStart[i] + lengthOfPencils[i]; ++id) {
+            if (*id && targetCells.contains(*id)) {
+               binsCells[i].insert(*id);
+            }
+         }
+      }
+
+      // Super ugly!
+      // TODO If anyone knows how to do this more efficiently feel free to fix it
+      std::set<uint> binsToDelete;
+      do {
+         binsToDelete.clear();
+         for (auto bin1 = binsCells.begin(); bin1 != binsCells.end(); ++bin1) {
+            if (binsToDelete.contains(bin1->first)) {
+               continue;
+            }
+
+            auto& cells1 = bin1->second;
+            for (auto bin2 = binsCells.begin(); bin2 != binsCells.end(); ++bin2) {
+               if (bin1 == bin2 || binsToDelete.contains(bin2->first)) {
+                  continue;
+               }
+
+               // Check for overlapping cells
+               for (auto cell : bin2->second) {
+                  if (cells1.contains(cell)) {
```

Comment:
and this is target cells of bin2. Thus, there isn't a check to see if there's an overlap of non-target source cells with the target cells of another pencil. Unless it happens somewhere else and I'm missing it?

Reply:
UP?

Reply:
ok, so the check is there, it was just too obfuscated for me to see it. Let's fix those namings and add descriptive comments please! How about doxygen-style? :)